Update filter documentation now that #351 is resolved. (#533)
kroxylicious/kroxylicious: docs/custom-filters.adoc

As seen above, filter methods (`onXyz[Request|Response]`) must return a `CompletionStage<FilterResult>` object.
It is the job of `FilterResult` to convey what message is to forwarded to the next filter in the chain (or broker
/client if at the chain's beginning or end).  It is also used to carry instructions such as indicating that the
connection must be closed, or a message dropped.

If the filter returns a `CompletionStage` that is already completed normally, Kroxylicious will immediately perform
the action described by the `FilterResult`.

The filter may return a `CompletionStage` that is not yet completed. When this happens, Kroxylicious will pause
reading from the downstream (the Client writes will eventually block), and it begins to queue up in-flight
requests/responses arriving at the filter.  This is done so that message order is maintained.  Once the
`CompletionStage` completes, the action described  by the `FilterResult` is performed, reading from the downstream
resumes and any queued up requests/responses are processed.

IMPORTANT: The pausing of reads from the downstream is a relatively costly operation.  To maintain optimal performance

> filter implementations should minimise the occasions on which not yet completed `CompletionStage` are returned.

Reviewer: nit 
```suggestion
filter implementations should minimise the occasions on which an incomplete `CompletionStage` is returned.
```

[source,java]
----
public class FetchFilter implements FetchRequestFilter {
    public static final short METADATA_VERSION_SUPPORTING_TOPIC_IDS = (short) 12;

    @Override
    public CompletionStage<RequestFilterResult> onFetchRequest(ApiKeys apiKey,
                                                               RequestHeaderData header,
                                                               FetchRequestData request,
                                                               KrpcFilterContext context) {
        MetadataRequestData metadataRequest = new MetadataRequestData(); //<1>
        var topic = new MetadataRequestData.MetadataRequestTopic();
        topic.setTopicId(Uuid.randomUuid());
        metadataRequest.topics().add(topic);
        short version = METADATA_VERSION_SUPPORTING_TOPIC_IDS;
        CompletionStage<MetadataResponseData> stage = context.sendRequest(version, request); //<2>
        return stage.thenApply(metadataResponse -> mutateFetchRequest(metadataResponse, request)) // <3>
                    .thenCompose(mutatedFetchRequest -> context.forwardRequest(header, mutatedFetchRequest)); // <4>
    }
}
----

> <1> We construct a new request object.  This will be the asynchronous request.

Reviewer: ```suggestion
<1> We construct a new request object.  This will be forwarded asynchronously.
```

<2> We send the request towards the Cluster, specifying the api version to use and obtaining a CompletionStage for the
response.

> <3> We use CompletionStage chaining to mutate the filtered fetch request using the asynchronous response.

Reviewer: ```suggestion
<3> We use CompletionStage chaining to mutate the filtered fetch request using the response from the request sent at <1>.
```

Author: https://github.com/asciidoctor/asciidoctor/issues/1077

There isn't a supported way to reference the callouts (so there is no fancy hyperlinking), but what you have done seems to render in a reasonable way.
